server/schedulers/scatter_range.go

```diff
@@ -15,48 +15,146 @@ package schedulers
 
 import (
 	"fmt"
-	"net/url"
+	"net/http"
+	"sync"
 
+	"github.com/gorilla/mux"
+	"github.com/pingcap/pd/pkg/apiutil"
+	"github.com/pingcap/pd/server/core"
 	"github.com/pingcap/pd/server/schedule"
 	"github.com/pingcap/pd/server/schedule/operator"
 	"github.com/pkg/errors"
+	"github.com/unrolled/render"
 )
 
 func init() {
-	schedule.RegisterScheduler("scatter-range", func(opController *schedule.OperatorController, args []string) (schedule.Scheduler, error) {
+	schedule.RegisterArgsToMapper("scatter-range", func(args []string) (schedule.ConfigMapper, error) {
 		if len(args) != 3 {
 			return nil, errors.New("should specify the range and the name")
 		}
-		startKey, err := url.QueryUnescape(args[0])
-		if err != nil {
-			return nil, err
+		mapper := make(schedule.ConfigMapper)
+		mapper["start-key"] = args[0]
+		mapper["end-key"] = args[1]
+		mapper["range-name"] = args[2]
+		return mapper, nil
+	})
+
+	schedule.RegisterScheduler("scatter-range", func(opController *schedule.OperatorController, storage *core.Storage, mapper schedule.ConfigMapper) (schedule.Scheduler, error) {
+		if len(mapper) != 3 {
+			return nil, errors.New("should specify the range and the name")
+		}
+		rangeName := mapper["range-name"].(string)
+		if len(rangeName) == 0 {
+			return nil, errors.New("the range name is invalid")
+		}
+		config := &scatterRangeSchedulerConf{
+			mu:        &sync.RWMutex{},
+			storage:   storage,
+			StartKey:  mapper["start-key"].(string),
+			EndKey:    mapper["end-key"].(string),
+			RangeName: rangeName,
 		}
-		endKey, err := url.QueryUnescape(args[1])
+
+		// persist config firstly
+		err := storage.SaveScheduleConfig(config.getScheduleName(), config)
 		if err != nil {
 			return nil, err
 		}
-		name := args[2]
-		return newScatterRangeScheduler(opController, []string{startKey, endKey, name}), nil
+		return newScatterRangeScheduler(opController, storage, config), nil
 	})
 }
 
+const scatterRangeScheduleType = "scatter-range"
+
+type scatterRangeSchedulerConf struct {
+	mu        *sync.RWMutex
+	storage   *core.Storage
+	RangeName string `json:"range-name"`
+	StartKey  string `json:"start-key"`
+	EndKey    string `json:"end-key"`
+}
+
+func (conf *scatterRangeSchedulerConf) BuildWithArgs(args []string) error {
+	if len(args) != 3 {
+		return errors.New("scatter range need 3 arguments to setup config")
+	}
+	conf.mu.Lock()
+	defer conf.mu.Unlock()
+
+	conf.RangeName = args[0]
+	conf.StartKey = args[1]
+	conf.EndKey = args[2]
+	return nil
+}
+
+func (conf *scatterRangeSchedulerConf) Clone() *scatterRangeSchedulerConf {
+	conf.mu.RLock()
+	defer conf.mu.RUnlock()
+	cpStartkey := string([]byte(conf.StartKey))
```

Comment:
Why?

Reply:
copy the string data.

Reply:
no need to copy string. Strings are readonly.